tla/counterfeit-cycles/counterfeit_cycles.tla

```diff
@@ -0,0 +1,295 @@
+    /\ LET msg == Head(ledger.msgs[s]) IN
+        /\ msg.type = TRANSFER
+        /\ msg.amount <= ledger.balances[msg.from]
+        /\ ledger' = [
+            ledger EXCEPT
+            \* Transaction is valid. Update ledger balances.
+            !["balances"][msg.from] = @ - msg.amount,
+            !["balances"][msg.to] = @ + msg.amount,
+
+            \* Remove msg from queue.
+            !["msgs"][s] = Tail(@)]
+        /\ subnetMsgs'= subnetMsgs \union {[
+            id |-> msg.id,
+            type |-> APPROVE,
+            from |-> msg.from,
+            to |-> msg.to,
+            amount |-> msg.amount
+            ]}
+        /\ UNCHANGED<<subnets, numDishonestActions, numTransfers>>
+
+Idle ==
+    \* Do nothing.
+    \* Once the maximum number of transfers is reached, the system is idle.
+    \* This is added to prevent TLA from thinking that it's stuck in a deadlock.
+    /\ (numTransfers >= MAX_TRANSFERS \/ numDishonestActions >= MAX_DISHONEST_TRANSFERS)
+    /\ UNCHANGED<<ledger, subnets, subnetMsgs, numDishonestActions, numTransfers>>
+
+Next ==
+  \/ SubnetSendTransfer
+  \/ SubnetReceiveTransfer
+  \/ SubnetReceiveApprove
+  \/ SubnetDishonestSendTransfer
+  \/ LedgerReceiveTransferMessage
+  \/ LedgerPoll
+  \/ Idle
+
+BalancesNonNegative ==
+  (*************************************************************************)
+  (* Invariant to ensure that all subnet balances are non-negative.        *)
+  (*************************************************************************)
+  \A s \in SUBNETS: subnets[s].balance >= 0
+
+RECURSIVE RecordToSeq(_,_)
+RecordToSeq(f, ks) ==
+  \* Helper method to transform a record's values into a sequence.
+  IF ks = {} THEN << >>
+  ELSE << f[CHOOSE k \in ks: TRUE] >> \o RecordToSeq(f, ks \ {CHOOSE k \in ks: TRUE})
+
+RECURSIVE SumSeq(_)
+SumSeq(s) ==
+  \* Helper method to sum a sequence
+  IF Len(s) = 0 THEN 0
+  ELSE Head(s) + SumSeq(Tail(s))
+
+RECURSIVE SumBalance(_)
+SumBalance(s) ==
+  \* Helper method to sum a sequence
+  IF Len(s) = 0 THEN 0
+  ELSE Head(s).balance + SumBalance(Tail(s))
+
+TotalSupplyInLedgerRemainsConstant ==
+    SumSeq(RecordToSeq(ledger.balances, SUBNETS)) = TOTAL_SUPPLY
+
+NoFakeCyclesAreCreated ==
+  \* The total supply of cycles that each subnet thinks it has is capped.
+  /\ SumBalance(RecordToSeq(subnets, SUBNETS)) <= TOTAL_SUPPLY
+
+SubnetNeverHasMoreBalanceThanLedger ==
+  \* The subnet's true balance is at most the balance stored in the ledger
+  \* for that subnet.
+  \A s \in SUBNETS: subnets[s].balance <= ledger.balances[s]
+
+\* TODO: is this invariant evaluated ever?
+LedgerAndSubnetBalancesMatchAfterMsgProcessing ==
+  \* Ledger and subnets have the same balances whenever all messages are
+  \* processed.
+  /\ (\A s \in SUBNETS: Len(ledger.msgs[s]) = 0)
+    /\ (\A s \in SUBNETS: Len(subnets[s].msgsToLedger) = 0)
+    /\ Cardinality(subnetMsgs) = 0
+  => \A sn \in SUBNETS: ledger.balances[sn] = subnets[sn].balance
+
```

Comment:
I'd add a few checks such as
```suggestion
SanityCanGetApprovals == \A msg \in subnetMsgs: msg.type # APPROVE
SanityNoMoreThanOneMessage = Cardinality(subnetMsgs) < 2
```
And run the model to make sure that they are violated.

Reply:
I don't think that there are such tests. I'm personally a bit scared of the TLC code base :) but https://groups.google.com/g/tlaplus might be a place to discuss